### PR TITLE
MRG, ENH: Adding a function to import EEGLAB ICA.

### LINKED ISSUE #6917: Import ICA results form EEGLAB
#### Describe the problem
EEG .set files can contain ICA results. Related data are populated in fields icaweights, icasphere, and icawinv as far as I know. Although users can build mne.preprocessing.ICA objects by hand from these fields, as proposed in [here](https://mail.nmr.mgh.harvard.edu/pipermail//mne_analysis/2018-March/004781.html), this requires a good knowledge of the internals of MNE to deal correctly with pre-whitening and similar preprocessing steps such that the results obtained in EEGLAB are reproducible in MNE. It is also error-prone.

#### Describe your solution
To add an import function to MNE to import ICA components from EEGLAB .set files in a transparent manner.

### PR DESCRIPTION
**Pull request by christian-oreilly**
#### Reference issue
Fixes #6917.

#### What does this implement/fix?
Adding a function to import ica from EEGLAB .set files.

Todo:

- [x] Fix shapes of components using SVD

### CODE REVIEW

File: mne/preprocessing/ica.py [hunk 1]
```diff
@@ -2624,3 +2638,39 @@ def corrmap(icas, template, threshold="auto", label=None, ch_type="eeg",
         return template_fig, labelled_ics
     else:
         return None
+
+
+def read_eeglab_ica(file_name):
```
Comment (cbrnr) [inline mne/preprocessing/ica.py]: Since we already have `read_ica` I suggest to rename this function to `read_ica_eeglab` (in line with `read_raw_*` conventions).

File: mne/preprocessing/ica.py [hunk 2]
```diff
@@ -58,9 +60,8 @@
 from .ctps_ import ctps
 from ..io.pick import channel_type, pick_channels_regexp
 
-
 __all__ = ('ICA', 'ica_find_ecg_events', 'ica_find_eog_events',
-           'get_score_funcs', 'read_ica', 'run_ica')
+           'get_score_funcs', 'read_ica', 'run_ica', 'read_eeglab_ica')
```
Comment (agramfort) [inline mne/preprocessing/ica.py]: It should be called read_ica_eeglab

File: mne/preprocessing/ica.py [hunk 3]
```diff
@@ -2624,3 +2638,39 @@ def corrmap(icas, template, threshold="auto", label=None, ch_type="eeg",
         return template_fig, labelled_ics
     else:
         return None
+
+
+def read_eeglab_ica(file_name):
+    """Load ica information saved in an EEGLAB .set file.
```
Comment (agramfort) [inline mne/preprocessing/ica.py]:
```suggestion
    """Load ICA information saved in an EEGLAB .set file.
```

File: mne/preprocessing/ica.py [hunk 4]
```diff
@@ -2624,3 +2638,39 @@ def corrmap(icas, template, threshold="auto", label=None, ch_type="eeg",
...
+
+    Parameters
+    ----------
+    file_name : Complete path to a .set EEGLAB file that contains an ica.
+
+    Returns
+    -------
+    ica : A mne.preprocessing.ICA object based on the information
```
Comment (agramfort) [inline mne/preprocessing/ica.py]: Please  check how we format docstrings

File: mne/preprocessing/ica.py [hunk 5]
```diff
@@ -1486,21 +1488,33 @@ def _pick_sources(self, data, include, exclude):
...
-        unmixing = np.dot(unmixing, self.pca_components_[:_n_pca_comp])
+        if (self.method == 'imported_eeglab' and
+           np.diff(self.unmixing_matrix_.shape)[0]):
+            # When \PCA reduction is used in EEGLAB, runica returns
+            # weights= weights*sphere*eigenvectors(:,1:ncomps)';
+            # sphere = eye(urchans);
+            # such that we cannot separate the unmixing matrix and
+            # the PCA components. For that reason, the composite
+            # unmixing matrix takes a rectangular shape.
+            unmixing = self.unmixing_matrix_
```
Comment (agramfort) [inline mne/preprocessing/ica.py]: I am surprised there is no dc offset involved.

File: mne/preprocessing/tests/test_ica.py [hunk 6]
```diff
@@ -1070,4 +1071,35 @@ def test_ica_eeg():
...
+                                                    comp_info["label"])
+                            if label in mark_to_remove]
+        ind_comp_to_drop = np.unique(np.concatenate(ind_comp_to_drop))
+    else:
+        ind_comp_to_drop = np.where(comp_info["flags"])[0]
+
+    ica = read_eeglab_ica(file_name)
+    raw_cleaned = ica.apply(raw.copy(), exclude=ind_comp_to_drop)
+
+    assert(np.all(np.isclose(raw_cleaned_matlab.get_data(),
```
Comment (agramfort) [inline mne/preprocessing/tests/test_ica.py]: Use numpy assert allclose

File: mne/preprocessing/ica.py (hunk 1 above)
Comment (christian-oreilly) [reply mne/preprocessing/ica.py]: Sure, make sense.

File: mne/preprocessing/ica.py (hunk 5 above)
Comment (christian-oreilly) [reply mne/preprocessing/ica.py]: As far as I know, EEGLAB does not export the dc offset. Anyway, the dc offset is just something that should be removed so that there is no offset. Maybe I am missing a point about this DC offset, but it does not make much sense to apply a DC offset for an ICA computed on some data to another dataset which is most likely to have another offset. The offset just need to be canceled out by .apply(), no?

File: mne/preprocessing/ica.py (hunk 2 above)
Comment (christian-oreilly) [reply mne/preprocessing/ica.py]: Corrected.

File: mne/preprocessing/ica.py (hunk 4 above)
Comment (christian-oreilly) [reply mne/preprocessing/ica.py]: I guess you are referring to the argument types that were missing. This has been corrected.

File: mne/preprocessing/tests/test_ica.py (hunk 6 above)
Comment (christian-oreilly) [reply mne/preprocessing/tests/test_ica.py]: Corrected.

**Comment by christian-oreilly** [comment]
Note: I made corrections directly through GitHub interfaces so it made a bunch of small commits, but I take for granted that the person that will merge this PR will "squash and merge".

**Comment by christian-oreilly** [comment]
@larsoner Do you know what could cause the Azure Pipelines to throw this kind of errors?
```
E           AssertionError: 
E           Found new public members missing from doc/python_reference.rst:
E           * read_ica_eeglab (mne.preprocessing.ica.read_ica_eeglab)
```
Do I need to add some doc somewhere besides the docstrings in the newly added function?

**Comment by cbrnr** [comment]
@christian-oreilly you need to add that function to `doc/python_reference.rst` because it is public.

File: mne/preprocessing/ica.py [hunk 7]
```diff
@@ -2624,3 +2638,39 @@ def corrmap(icas, template, threshold="auto", label=None, ch_type="eeg",
...
+    """Load ICA information saved in an EEGLAB .set file.
+
+    Parameters
+    ----------
+    file_name : str
+        Complete path to a .set EEGLAB file that contains an ICA.
+
+    Returns
+    -------
+    ica : mne.preprocessing.ICA object
```
Comment (agramfort) [inline mne/preprocessing/ica.py]:
```suggestion
    ica : instance of ICA
```

File: mne/preprocessing/ica.py [hunk 8]
```diff
@@ -2624,3 +2638,39 @@ def corrmap(icas, template, threshold="auto", label=None, ch_type="eeg",
         return template_fig, labelled_ics
     else:
         return None
+
+
+def read_ica_eeglab(file_name):
```
Comment (agramfort) [inline mne/preprocessing/ica.py]: input are commonly called `fname`

File: mne/preprocessing/ica.py [hunk 9]
```diff
@@ -2624,3 +2638,39 @@ def corrmap(icas, template, threshold="auto", label=None, ch_type="eeg",
...
+
+    Parameters
+    ----------
+    file_name : str
+        Complete path to a .set EEGLAB file that contains an ICA.
+
+    Returns
+    -------
+    ica : mne.preprocessing.ICA object
+        An ICA object based on the information contained in file file_name.
```
Comment (agramfort) [inline mne/preprocessing/ica.py]:
```suggestion
        An ICA object based on the information contained in the input file.
```

File: mne/preprocessing/tests/test_ica.py [hunk 10]
```diff
@@ -1070,4 +1071,34 @@ def test_ica_eeg():
...
+                                                    comp_info["label"])
+                            if label in mark_to_remove]
+        ind_comp_to_drop = np.unique(np.concatenate(ind_comp_to_drop))
+    else:
+        ind_comp_to_drop = np.where(comp_info["flags"])[0]
+
+    ica = read_ica_eeglab(file_name)
+    raw_cleaned = ica.apply(raw.copy(), exclude=ind_comp_to_drop)
+
+    assert np.allclose(raw_cleaned_matlab.get_data(), raw_cleaned.get_data())
```
Comment (agramfort) [inline mne/preprocessing/tests/test_ica.py]: use `from numpy.testing import assert_allclose`

Comment (christian-oreilly) [reply mne/preprocessing/tests/test_ica.py]: Done.

File: mne/preprocessing/ica.py (hunk 8 above)
Comment (christian-oreilly) [reply mne/preprocessing/ica.py]: file_name changed to fname in read_ica_eeglab and test_read_ica_eeglab.

File: mne/preprocessing/ica.py [hunk 11]
```diff
@@ -1486,21 +1488,33 @@ def _pick_sources(self, data, include, exclude):
...
 
         if _n_pca_comp > n_components:
             sel_keep = np.concatenate(
                 (sel_keep, range(n_components, _n_pca_comp)))
 
-        proj_mat = np.dot(mixing[:, sel_keep], unmixing[sel_keep, :])
+        proj_mat = mixing[:, sel_keep] @ unmixing[sel_keep, :]
 
-        data = np.dot(proj_mat, data)
+        data = proj_mat @ data
```
Comment (jona-sassenhagen) [inline mne/preprocessing/ica.py]: Love all of the `@`s

File: mne/preprocessing/ica.py [hunk 12]
```diff
@@ -2624,3 +2638,39 @@ def corrmap(icas, template, threshold="auto", label=None, ch_type="eeg",
...
+    fname : str
+        Complete path to a .set EEGLAB file that contains an ICA.
+
+    Returns
+    -------
+    ica : instance of ICA
+        An ICA object based on the information contained in the input file.
+    """
+    eeg = _check_load_mat(fname, None)
+    info, _, _ = _get_info(eeg)
```
Comment (jona-sassenhagen) [inline mne/preprocessing/ica.py]: or

```suggestion
    info = _get_info(eeg)[0]
```

... which would make it robust against changing the number of return parameters and not lead to one of the `_`s lingering in memory.

File: mne/preprocessing/ica.py (hunk 11 above)
Comment (christian-oreilly) [reply mne/preprocessing/ica.py]: Yeah.... I read that it was now the recommended way. And it is definitly more readable. So I though why not update this part of code on the way... :)

**Comment by larsoner** [comment]
Diff looks remarkably clean, I expected it to be more painful. But I'm guessing the painful part was getting the math right :)

@christian-oreilly can you update `doc/changes/latest.inc` to include this? Then +1 for merge from me. @cbrnr are you also happy?

**Comment by cbrnr** [comment]
I didn't check the math so I trust that it works based on the test file. If we notice a bug later, we can always correct it then. So 👍 for merge.

File: mne/preprocessing/tests/test_ica.py [hunk 13]
```diff
@@ -1070,4 +1071,35 @@ def test_ica_eeg():
...
+        ind_comp_to_drop = [np.where(flags)[0]
+                            for flags, label in zip(comp_info["flags"],
+                                                    comp_info["label"])
+                            if label in mark_to_remove]
+        ind_comp_to_drop = np.unique(np.concatenate(ind_comp_to_drop))
+    else:
+        ind_comp_to_drop = np.where(comp_info["flags"])[0]
+
+    ica = read_ica_eeglab(fname)
+    raw_cleaned = ica.apply(raw.copy(), exclude=ind_comp_to_drop)
```
Comment (agramfort) [inline mne/preprocessing/tests/test_ica.py]: @christian-oreilly are the bad ica components marked in the set files? if so, why not reading it and setting on the read the ica.exclude attribute?

File: mne/preprocessing/ica.py (hunk 5 above)
Comment (agramfort) [reply mne/preprocessing/ica.py]: ok then EEGLAB does not store the `pca_mean_` and recomputes or not the centering on the new data on which it's applied.

File: mne/preprocessing/ica.py [hunk 14]
```diff
@@ -1486,21 +1488,33 @@ def _pick_sources(self, data, include, exclude):
...
-        unmixing[:n_components, :n_components] = self.unmixing_matrix_
-        unmixing = np.dot(unmixing, self.pca_components_[:_n_pca_comp])
+        if (self.method == 'imported_eeglab' and
+           np.diff(self.unmixing_matrix_.shape)[0]):
+            # When \PCA reduction is used in EEGLAB, runica returns
+            # weights= weights*sphere*eigenvectors(:,1:ncomps)';
+            # sphere = eye(urchans);
+            # such that we cannot separate the unmixing matrix and
+            # the PCA components. For that reason, the composite
+            # unmixing matrix takes a rectangular shape.
```
Comment (agramfort) [inline mne/preprocessing/ica.py]: this makes the current docstring not correct all the time. Can you add a sentence to say this in the docstring that documents attributes?

File: mne/preprocessing/tests/test_ica.py (hunk 13 above)
Comment (christian-oreilly) [reply mne/preprocessing/tests/test_ica.py]: These marks are from an extension to EEGLAB named [vised-marks](https://github.com/BUCANL/Vised-Marks) which allows to mark epochs, channels and components. These are marked with a label that is defined by the user (e.g., manual, eog, ecg, ...). Which components should be removed depends on the intended usage and the user decision. In this specific case, when I created the test file in Matlab, I marked three channels using the string "manual" and saved in EEGLAB a version of the file where these components were rejected to simulate an artifact rejection preprocessing. Thus, for comparison in the test, I do the same thing in MNE. I don't think that it can be generalized so that components are marked in ica.exclude when importing the ICA object.

File: mne/preprocessing/ica.py (hunk 14 above)
Comment (christian-oreilly) [reply mne/preprocessing/ica.py]: I've adjusted the docstrings accordingly.

**Comment by christian-oreilly** [comment]
Thanks @larsoner. I tried to keep it clean. 

Yes, the math and understanding what is call what in EEGLAB versus MNE and how to do acceptable correspondances took most of the time.

I tested it with three datasets with ICA performed under different context. It might still break down when users will try it on ICA parametrized in ways I did not tested. Since the .set files are juste normal matlab files, users or different EEGLAB extension can modified them in ways I did not forseen. Also, I am also not terribly satisfied from rounding errors but the operations seems right as far as I can tell so I would attribute it matrix inversions or rounding when data are dumped to disk in .set and read back

Aside from the official tests, I did not see issues so far on the 100 hd-EEG recordings I used it in for database that I am currently analysing.

**Comment by agramfort** [comment]
just thinking out loud but could you compute an SVD of the mixing matrix to
obtain by the MNE pca_components and the mixing with the same shapes. It
would simplify API documentation.

>

**Comment by christian-oreilly** [comment]
@agramfort Hmm... I did not think of it. I am not super proficient with ICA/SVD math, but we would want to decompose `weights*sphere*eigenvectors(:,1:ncomps)` into `weights*sphere` and `eigenvectors(:,1:ncomps)`. The term `eigenvectors(:,1:ncomps)` corresponds well to S in the SVD decomposition of `M=weights*sphere*eigenvectors(:,1:ncomps)`  into `M=USV*` (i.e., `S` and `eigenvectors(:,1:ncomps)` have the same definition: MxN diagonal matrix of eigenvalues). Then, I am not clear what to do with the `V*`. We would need it to be the identity matrix. So, it would work if `MV=USV*V=U'S'V'*` is mathematically such that SVD gives us the decomposition for which `V'*=1` and `S'=S`. I am not sure if we have such guaranty. If we did, then we would get `weights*sphere=U'*MV`. I can verify if it checks out. It would definitly make the code much cleaner, not just the doc but also the the consistency of the values taken by the ICA attributes.

**Comment by agramfort** [comment]
rendering of docstring look wrong https://16261-1301584-gh.circle-artifacts.com/0/dev/generated/mne.preprocessing.ICA.html#mne.preprocessing.ICA not sure if it's your fault @christian-oreilly 

@larsoner any idea?

btw I find the layout of the parameters from sklearn much more readable eg:

https://scikit-learn.org/stable/modules/generated/sklearn.cluster.AffinityPropagation.html#sklearn.cluster.AffinityPropagation

basically I would not expect the param description to be on the same line at the name + type that should stand out.

**Comment by christian-oreilly** [comment]
>  rendering of docstring look wrong https://16261-1301584-gh.circle-artifacts.com/0/dev/generated/mne.preprocessing.ICA.html#mne.preprocessing.ICA not sure if it's your fault @christian-oreilly

@agramfort Yes, it is probably my fault. CI checks worked before the last commit I made where I changed the docstrings.

> just thinking out loud but could you compute an SVD of the mixing matrix to
> obtain by the MNE pca_components and the mixing with the same shapes. It
> would simplify API documentation.

I sure would. I had a look at it a few days ago, without much luck. I will look into it again when I get some time. Right now, I shifted to other duties.

**Comment by agramfort** [comment]
@christian-oreilly do you have time to make it green again? i can live with the current design

**Comment by christian-oreilly** [comment]
@larsoner No, I did not have much time to take a look at it again. Last time I did, I had the feeling that it might not be possible to do it cleanly. From memory, the issue was boiling down to the fact that we would want to decompose M=X@Y into U@S@V', with X, U and V' being rectangular and S and Y being square. So my problem was about how you get rid of the V' so that you can get the correspondences U=X and S=Y. I tried to transform data such that V' would become the identity matrix but without success. Feel free to have a look at it. Maybe a fresh set of eyes on this problem would be beneficial.

@agramfort I just looked at the CI issues more closely. It does not look like it was related to my last commit in the end. It was a socket error when running test_fetch_parcellations. No changes I made should have any effect on that test. Maybe just a transient network failure. I re-triggered the tests to check if they pass now.

**Comment by christian-oreilly** [comment]
@larsoner Is this PR still on your radar? You were supposed to have a look at it around end of October.

**Comment by larsoner** [comment]
Forgot about this, will try to look this week

**Comment by larsoner** [comment]
@christian-oreilly okay for me to rebase and force push? I think I got this working but it would be good if you could double-check.

File: mne/preprocessing/ica.py [hunk 15]
```diff
@@ -2624,3 +2640,53 @@ def corrmap(icas, template, threshold="auto", label=None, ch_type="eeg",
...
+        # weights= weights*sphere*eigenvectors(:,1:ncomps)';
+        # sphere = eye(urchans), so let's use SVD to get our square
+        # weights matrix (u * s) and our PCA vectors (v) back
+        u, s, v = _safe_svd(eeg.icaweights, full_matrices=False)
+        ica.unmixing_matrix_ = u * s
+        ica.pca_components_ = v
+    else:
+        ica.unmixing_matrix_ = eeg.icaweights
+        ica.pca_components_ = eeg.icasphere
+    ica.mixing_matrix_ = linalg.pinv(ica.unmixing_matrix_)
```
Comment (larsoner) [inline mne/preprocessing/ica.py]: @christian-oreilly this part I'm not sure about. We no longer dot with `eeg.icawinv` after the `linalg.pinv` here, which you had. But in MNE we should have that the mixing matrix is just the pinv of the unmixing matrix:

https://github.com/mne-tools/mne-python/blob/master/mne/preprocessing/ica.py#L2138

Comment (christian-oreilly) [reply mne/preprocessing/ica.py]: I don't remember the exact details here because it was a bit complicated and its been a few month I looked at it now, but in a nutshell, that is what I remember from my conclusion: As far as I understand, there are no exact correspondence between the variables exposed by EEGLAB and MNE for ICA. Futher, after looking for whether EEGLAB exposed variables could be transformed in what MNE is expecting (e.g., through SVD), I was inclined to believed that there were no easy way to do it. So, as far as I know, we will not manage to have something semantically (with respect to the definition of ica.unmixing_matrix and ica.pca_components_ in MNE) perfect here. Given this limitation, my efforts were directed at having an import function that gives numerically the correct output when the ICA is applied in MNE versus when applied in EEGLAB; I validated this by running tests. One of these tests is part of this PR.

**Comment by christian-oreilly** [comment]
@larsoner Sure, should be OK to rebase and force push. What do you mean by double-checking this exactly? Pulling the last state of the PR and check if the tests runs on my side?

**Comment by larsoner** [comment]
Whatever tests you had implemented here passed on my machine. I added some assertions that shapes and such match what the MNE objects were, so hopefully those are correct now, too.

But yes please pull and try, and check whatever else you checked originally to make sure things worked. If it's just the test that's in the test suite that's fine.

If you want to look at the code and shapes, etc. again to see if what I did overcame the limitations you had before that would be great, but again if we can trust your tests then we should be okay.

**Comment by larsoner** [comment]
> So, as far as I know, we will not manage to have something semantically (with respect to the definition of ica.unmixing_matrix and ica.pca_components_ in MNE) perfect here

I added a `_assert_ica_attributes` class that hopefully checks to make sure our PCA / unmixing / mixing matrices meet the definitions that MNE wants. But @christian-oreilly it would be good if you could check and see if your workflows using this function still work, or at least some basic use cases do. For example, if you could check to see that some component waveforms look the same in EEGLAB and MNE that would be very helpful.

Otherwise this should be good for review/merge now. Tests pass locally for me and the code has been squashed to by-author commits, so a rebase+merge rather than squash+merge should be done here if possible.

**Comment by christian-oreilly** [comment]
@larsoner The test I added compare the signals transformed from applying the EEGLAB ICA to those obtained by applying the same ICA imported and applied to the signals in MNE. So I think it is a rather complete test that does verify that the waveform match between EEGLAB and MNE. I am in the middle of refactoring lots of code right now so I need to finish this refactoring before I can rerun my pipeline and see if everything works on my other uses cases. If you don't mind waiting something like a week before merging this PR, I can finish this refactoring, rerun the refactored pipeline with MNE patch with this PR and confirm that everything works fine on my end.

**Comment by larsoner** [comment]
Sure, should be fine to wait

File: mne/preprocessing/ica.py [hunk 16]
```diff
@@ -2680,8 +2672,18 @@ def read_ica_eeglab(fname):
     ica.pre_whitener_ = np.ones((len(eeg.icachansind), 1))
     ica.pca_mean_ = np.zeros(len(eeg.icachansind))
 
-    ica.unmixing_matrix_ = eeg.icaweights
-    ica.pca_components_ = eeg.icasphere
-    ica.mixing_matrix_ = np.linalg.pinv(ica.pca_components_) @ eeg.icawinv
-
+    n_ch = len(ica.ch_names)
+    assert eeg.icaweights.shape == (n_components, n_ch)
```
Comment (christian-oreilly) [inline mne/preprocessing/ica.py]: @larsoner  I cannot add this directly as a suggestion that you can commit because it is not on lines part of this commit, but we would have to change line 2667 from "ica.ch_names = info["ch_names"]" to something like "ica.ch_names = np.array(info["ch_names"])[eeg["icachansind"].astype(int)-1].tolist()". EEGLAB ICA can be computed on a subset of channels (indicated by the icachansind field) and when they are, your assert on line 2676 fails. With this change, it passes.

**Comment by larsoner** [comment]
@christian-oreilly does it work now?

**Comment by larsoner** [comment]
(forgot the `-1`, force-pushed the fix)

**Comment by larsoner** [comment]
@christian-oreilly can you try one more time? If all is well then we can merge

**Comment by christian-oreilly** [comment]
Yes, sorry for the delay. What I tested worked well. I think it can be merged. Thanks.

**Comment by larsoner** [comment]
@agramfort feel free to merge. Would be good to get a set of eyes on the additional checks that I put in place

**Comment by agramfort** [comment]
thx @christian-oreilly and @larsoner !

**Comment by dede3de** [comment]
Is there a documentation for mne.preprocessing.read_ica_eeglab() to understand the input to insert?
Also, can we use this function to add the projections to the EEG data that we need for source estimation analysis? Thanks!

**Comment by larsoner** [comment]
https://mne.tools/dev/generated/mne.preprocessing.read_ica_eeglab.html#mne.preprocessing.read_ica_eeglab

For ICA we do not account for the operation during source analysis IIRC

**Comment by christian-oreilly** [comment]
@dede3de You may want to have a look at `test_read_ica_eeglab()` in  `mne/preprocessing/tests/test_ica.py` if you did not already. It would give you an example of use.

File: mne/preprocessing/ica.py [hunk 17]
```diff
@@ -2680,8 +2672,18 @@ def read_ica_eeglab(fname):
     ica.pre_whitener_ = np.ones((len(eeg.icachansind), 1))
     ica.pca_mean_ = np.zeros(len(eeg.icachansind))
 
-    ica.unmixing_matrix_ = eeg.icaweights
```
Comment (christian-oreilly) [inline mne/preprocessing/ica.py]: I'm afraid this specific change introduced a bug. I don't think the code for `n_components < n_ch` works as expected. I'll have to check, but it is possible that `test_read_ica_eeglab` just validates the "else" part of this if, which corresponds more or less to the behavior of this function before the introduction of this if-else statement. I'll create an issue to describe this bug and submit a corresponding PR.
